src/yunohost/permission.py

```diff
@@ -349,180 +288,175 @@ def permission_add(operation_logger, app, permission, urls=None, default_allow=T
+    groups = user_group_list(full=True)["groups"]
+    permissions = user_permission_list(full=True)["permissions"]
+
+    for permission_name, permission_infos in permissions.items():
+
+        # These are the users currently allowed because there's an 'inheritPermission' object corresponding to it
+        currently_allowed_users = set(permission_infos["corresponding_users"])
+
+        # These are the users that should be allowed because they are member of a group that is allowed for this permission ...
+        should_be_allowed_users = set([user for group in permission_infos["allowed"] for user in groups[group]["members"]])
+
+        # Note that a LDAP operation with the same value that is in LDAP crash SLAP.
+        # So we need to check before each ldap operation that we really change something in LDAP
+        if currently_allowed_users == should_be_allowed_users:
+            # We're all good, this permission is already correctly synchronized !
             continue
-        inheritPermission = {'inheritPermission': user_permission, 'memberUid': uid_val}
-        if force:
-            if per['groupPermission']:
-                if not ldap.update('cn=%s,ou=permission' % per['cn'][0], {'groupPermission': []}):
-                    raise YunohostError('permission_update_failed_clear')
-                if not ldap.update('cn=%s,ou=permission' % per['cn'][0], {'groupPermission': per['groupPermission']}):
-                    raise YunohostError('permission_update_failed_populate')
-            if per['inheritPermission']:
-                if not ldap.update('cn=%s,ou=permission' % per['cn'][0], {'inheritPermission': []}):
-                    raise YunohostError('permission_update_failed_clear')
-            if user_permission:
-                if not ldap.update('cn=%s,ou=permission' % per['cn'][0], inheritPermission):
-                    raise YunohostError('permission_update_failed')
-        else:
-            if not ldap.update('cn=%s,ou=permission' % per['cn'][0], inheritPermission):
-                raise YunohostError('permission_update_failed')
-    logger.debug(m18n.n('permission_generated'))
+
+        new_inherited_perms = {'inheritPermission': ["uid=%s,ou=users,dc=yunohost,dc=org" % u for u in should_be_allowed_users],
+                               'memberUid': should_be_allowed_users}
+
+        # Commit the change with the new inherited stuff
+        try:
+            ldap.update('cn=%s,ou=permission' % permission_name, new_inherited_perms)
+        except Exception as e:
+            raise YunohostError('permission_update_failed', permission=permission_name, error=e)
+
+    logger.debug("The permission database has been resynchronized")
 
     app_ssowatconf()
 
     # Reload unscd, otherwise the group ain't propagated to the LDAP database
     os.system('nscd --invalidate=passwd')
     os.system('nscd --invalidate=group')
+
+
+def _normalize_url(url):
```

Comment:
Maybe we can move this helper in the domain module if we need it an other time somewhere.

Reply:
Well in fact in https://github.com/YunoHost/yunohost/pull/797 I removed it because I switched to relative URLs ...

Reply:
Ok, let's leave it here.